File: lib/csharp/src/Protocol/TTinyProtocol.cs
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 * Contains some contributions under the Thrift Software License.
 * Please see doc/old-thrift-license.txt in the Thrift distribution for
 * details.
 */

using System;
using System.Text;
using Thrift.Transport;
using System.Collections;
using System.IO;
using System.Collections.Generic;

namespace Thrift.Protocol
{
    public class TTinyProtocol : TCompactProtocol
    {
        #region CompactProtocol Factory

        public new class Factory : TProtocolFactory
        {
            public Factory() { }

            public TProtocol GetProtocol(TTransport trans)
            {
                return new TTinyProtocol(trans);
            }
        }

        #endregion

        public TTinyProtocol(TTransport trans) : base(trans) {}

		private string[] stringArray;
		private List<string> stringList;
		private Dictionary<string, int> stringDict;
		private int structDepth;
		private TTransport realTrans;
		private MemoryStream realStream;

		public override void WriteStructBegin(TStruct strct)
		{
			if (structDepth == 0)
			{
				stringList = new List<string> ();
				stringDict = new Dictionary<string, int> ();
				realTrans = trans;
				realStream = new MemoryStream ();
				trans = new TStreamTransport (null, realStream);
			}
			structDepth++;

			base.WriteStructBegin (strct);
		}

		public override void WriteStructEnd()
		{
			base.WriteStructEnd ();

			structDepth--;
			if (structDepth == 0)
			{
				trans = realTrans;
				WriteVarint32 ((uint)stringList.Count);
				foreach (var s in stringList)
					WriteBinary (UTF8Encoding.UTF8.GetBytes (s));
				var fullBytes = realStream.GetBuffer(); // XXX(ek): shouldn't this be ToArray()?
				trans.Write(fullBytes, 0, fullBytes.Length);
			}
		}

        public override void WriteString(String str)
        {
			int slot;
			if (!stringDict.TryGetValue (str, out slot))
			{
				slot = stringList.Count;
				stringList.Add (str);
				stringDict [str] = slot;
			}

			WriteVarint32 ((uint)slot);
        }

        public override TStruct ReadStructBegin()
        {
			// Read the string header.
			if (structDepth == 0)
			{
				int count = (int)ReadVarint32();
				stringArray = new string[count];

				for (int i = 0; i < count; ++i)
				{
					var str = Encoding.UTF8.GetString(ReadBinary());
					stringArray [i] = str;
				}
			}
			structDepth++;

			return base.ReadStructBegin();
        }

        public override void ReadStructEnd()
        {
			base.ReadStructEnd ();
			structDepth--;
        }

        public override String ReadString()
        {
			int index = (int)ReadVarint32();
			return stringArray [index];
        }
    }
}

File: lib/java/src/org/apache/thrift/protocol/TTinyProtocol.java
package org.apache.thrift.protocol;

import org.apache.thrift.TException;
import org.apache.thrift.transport.TMemoryBuffer;
import org.apache.thrift.transport.TTransport;

import java.util.LinkedHashMap;

public class TTinyProtocol extends TCompactProtocol {

    public TTinyProtocol(TTransport tTransport) {
        super(tTransport);
    }

    int structDepth = 0;
    String[] stringArray;
    LinkedHashMap<String, Integer> stringDict;
    private TTransport realTrans;
    private TMemoryBuffer memoryTrans;

    @Override
    public void writeStructBegin(TStruct strct) throws TException {
        if (structDepth == 0) {
            stringDict = new LinkedHashMap<>();
            realTrans = trans_;
            memoryTrans = new TMemoryBuffer(32);
            trans_ = memoryTrans;
        }
        structDepth++;

        super.writeStructBegin(strct);
    }

    @Override
    public void writeStructEnd() throws TException {
        super.writeStructEnd();

        structDepth--;
        if (structDepth == 0) {
            trans_ = realTrans;
            writeVarint32((int) stringDict.size());
            for (String s : stringDict.keySet())
                super.writeString(s);
            trans_.write(memoryTrans.getArray(), 0, memoryTrans.length());
        }
    }

    @Override
    public void writeString(String str) throws TException {
        int slot;
        if (!stringDict.containsKey(str)) {
            slot = stringDict.size();
            stringDict.put(str, slot);
        } else {
            slot = stringDict.get(str);
        }

        writeVarint32(slot);
    }

    public TStruct readStructBegin() throws TException {
        // Read the string header.
        if (structDepth == 0) {
            int count = readVarint32();
            stringArray = new String[count];
            for (int i = 0; i < count; ++i) {
                stringArray[i] = super.readString();
            }
        }
        structDepth++;
        return super.readStructBegin();
    }

    public void readStructEnd() throws TException {
        super.readStructEnd();
        structDepth--;
    }

    public String readString() throws TException {
        int index = this.readVarint32();
        return stringArray[index];
    }

    private int readVarint32() throws TException {
        int result = 0;
        int shift = 0;
        if (this.trans_.getBytesRemainingInBuffer() >= 5) {
            byte[] var7 = this.trans_.getBuffer();
            int pos = this.trans_.getBufferPosition();
            int off = 0;

            while (true) {
                byte b1 = var7[pos + off];
                result |= (b1 & 127) << shift;
                if ((b1 & 128) != 128) {
                    this.trans_.consumeBuffer(off + 1);
                    break;
                }

                shift += 7;
                ++off;
            }
        } else {
            while (true) {
                byte b = this.readByte();
                result |= (b & 127) << shift;
                if ((b & 128) != 128) {
                    break;
                }

                shift += 7;
            }
        }

        return result;
    }

    /**
     * Temporary buffer used for various operations that would otherwise require a
     * small allocation.
     */
    private final byte[] temp = new byte[10];

    /**
     * Write an i32 as a varint. Results in 1-5 bytes on the wire.
     * TODO: make a permanent buffer like writeVarint64?
     */
    private void writeVarint32(int n) throws TException {
        int idx = 0;
        while (true) {
            if ((n & ~0x7F) == 0) {
                temp[idx++] = (byte) n;
                // writeByteDirect((byte)n);
                break;
                // return;
            } else {
                temp[idx++] = (byte) ((n & 0x7F) | 0x80);
                // writeByteDirect((byte)((n & 0x7F) | 0x80));
                n >>>= 7;
            }
        }
        trans_.write(temp, 0, idx);
    }

    public static class Factory implements TProtocolFactory {

        public TProtocol getProtocol(TTransport trans) {
            return new TTinyProtocol(trans);
        }
    }
}

